envoy/event/dispatcher.h

```diff
@@ -218,6 +218,19 @@ class Dispatcher : public DispatcherBase, public ScopeTracker {
                          Network::TransportSocketPtr&& transport_socket,
                          const Network::ConnectionSocket::OptionsSharedPtr& options) PURE;
 
+  /**
+   * Register an internal listener manager for this dispatcher.
+   */
+  virtual void
+  registerInternalListenerManager(Network::InternalListenerManager& internal_listener_manager) PURE;
+
+  /**
+   * @brief Get the Internal Listener Manager object.
+   *
+   * @return the registered internal istener manager or nullopt.
+   */
+  virtual Network::InternalListenerManagerOptRef getInternalListenerManager() PURE;
```

Comment:
Could we avoid use of the dispatcher interface as a registration mechanism for factories like this one?

Reply:
It seems there was slight agreement on this  so I stopped investing on bootstrap extension.

My opinion is that adding this interface to dispatcher is more straightforward, well, at the cost of adding getter/setter  

CC @yanavlasov

Reply:
I think we need a client connection factory extension point that connection pools and other things can use to create connections.  An eventual goal may be to completely remove the client connection creation methods from the dispatcher.  But this is something that @envoyproxy/senior-maintainers may want to chime in on.

Reply:
I see two arguments here,
the first is to store the internal listener registry, here I choose dispatcher over bootstrap extension(where we can add a threadlocal registery), because the life time of dispatcher and listener manager is cleaner. The dispatcher is guaranteed existing when an internal listener is ready to add and use. 

I think we can add the complexity in booting the server, and offer a hook for internal listener registry if we choose to store the internal listeners there. I feel we can do that after. 

the second is to remove the thin dispatcher::createClientConnection. Again we can do that along with this PR but I prefer to do it later when signed off.

Reply:
The current approach of adding the internal listener registry to the dispatcher does not generalize to the next extension that needs to hook into the code that creates client connections.  I recommend introducing a factory mechanism for client connections and transition to it, eventually deprecating the current Dispatcher::createClientConnection

Reply:
I agree with @antoniovicente that the extension behavior has leaked into core interfaces. II think it would be cleaner if the mechanism for finding the listener was completely encapsulated in the factory. For example when internal listener is created it is registering itself in the factory's thread local storage, where it can later be looked up for establishing client connections. In this way you can avoid needing to change the dispatcher interface and ConnectionHandlerImpl